sycl/doc/extensions/proposed/sycl_ext_intel_grf_size.asciidoc

```diff
@@ -0,0 +1,143 @@
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties]
+* link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
+  sycl_ext_oneapi_kernel_properties]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Backend support status
+
+This extension is currently implemented in {dpcpp} only for Intel GPU devices and
+only when using the Level Zero backend or OpenCL backend.
+Attempting to use this extension in kernels that run on other devices or
+backends may result in undefined behavior.  Be aware that the compiler
+is not able to issue a diagnostic to warn you if this happens.
+
+== Overview
+
+There are devices where the size of the general register file (GRF) used by a kernel is 
+configurable. Developers may want to change the GRF size based on their
+application. This extension adds the kernel property `grf_size` which provides a way
+to specify the GRF size.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_GRF_SIZE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== GRF Size Property
+
+The `grf_size` kernel property provides a way to specify the GRF size used by a kernel. 
+It accepts a single integer value. It is defined as follows:
+```c++
+namespace sycl::ext::intel::experimental {
+
+struct grf_size_key {
+  template <int Size>
+  using value_t = 
+      oneapi::experimental::property_value<grf_size_key, 
+                                           std::integral_constant<int, Size>>;
+};
+
+template <int Size>
+inline constexpr grf_size_key::value_t<Size> grf_size;
+
+} // namespace sycl::ext::intel::experimental
+```
+The supported values are as follows:
+[%header,cols="1,5"]
+|===
+|GPU |Supported Values
+| PVC | 128 (small register file), 256 (large register file)
```

Comment:
@gmlueck @bader @steffenlarsen I just remembered we forgot to cover a very important usecase: automatic register file selection. this means the backend uses heuristics to choose either small or large. we have users who want to use that as well.

how do you recommend covering that? i can think of two options:
1) add a new property `auto_grf_size` that is mutually exclusive with this one
2) add a `-1` or `0` value to the supported values

Reply:
I recommend adding a new property.  That's consistent with how we handled this for sub-groups: we have `sub_group_size<uint32_t>` for user-supplied sizes, and `sub_group_size_automatic` for compiler-chosen size (see [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc)).

`grf_size_automatic` would match the naming there.

Reply:
If you go with (1) we should replace existing integer values with some enum values (e.g. enum `grf_size` with values `small`, `large`, `auto`).

(2) option seems to be more future-proof. I mean there might be no changes in API required to support more/less integer values. I would probably add clarification about the meaning of the `Size` integer. If I get it correctly it's the number of registers in the file. Right?

Reply:
How would you implement the "auto" case?  I thought we plan to implement this property on top of the Level Zero online compiler flag `-ze-opt-large-register-file`.  That only provides two possibilities "large" and "small".  (Or, maybe it's "large" and "auto".  I'm not sure what the default is.)

Reply:
@gmlueck "auto" is actually already implemented in the `sycl::detail` version. There is the `-ze-intel-enable-auto-large-GRF-mode` flag to do it.

Based on @Pennycook's comment about `sub_group_size_automatic`, option 1 seems to be better for consistency. 

is there any agreement with @bader's comment to use enums, similar to the current `sycl::detail` version, as opposed to a number of registers?

Reply:
i also prefer the register number as it seems more scalable for possible future hardware, maybe there's some new hardware that supports 128, 256 and 512 or something, we'd have to add a new value to the enum to support all sizes.

Reply:
Should we change the type of `Size` template parameter to `unsigned int`? 
NOTE: Using an unsigned type can affect performance.

Reply:
Since we seem to agree to use a new property for the `auto` case, we shouldn't need any sentinel value like `-1`, so `unsigned int` should be fine if it improves performance.

Reply:
> i also prefer the register number as it seems more scalable for possible future hardware, maybe there's some new hardware that supports 128, 256 and 512 or something, we'd have to add a new value to the enum to support all sizes.

Agree. That's better than hardcoded enums

Reply:
> That flag isn't documented in the Level Zero spec. @jandres742 is -ze-intel-enable-auto-large-GRF-mode something official that we can rely on? Can it be added to the Level Zero spec?

@gmlueck : right, it is in the driver implemented, but not in the spec documented. Open issue in spec to track it https://github.com/oneapi-src/level-zero-spec/issues/148

Reply:
ok, i think there is decent consensus here, let me try to make an updated draft